### PR TITLE
message processing optimizations, multi-codec support

### PR DESCRIPTION
**Pull request by colhom**
Most of the changes in this PR focus on allowing Rviz to process frames as quickly as possible. This primarily entailed decoupling and parallelizing:
  * video encoding (we run two encoders per RViz, regular and keyed)
  * dbus back-and-forth
  * Ogre rendering

This PR also introduces multi-codec support, mainly by exposing an `h264_encoder` parameter.

### CODE REVIEW

**Review by ccpost — COMMENTED** [review]
@colhom In general please remember to request review or PRs will go unread.

Python bit LGTM.  @delaneyj is much more qualified than I am to give a 👍 to the RViz C++ parts.

**Review by delaneyj — CHANGES_REQUESTED** [review]
I love that you are using the signal/slot interface to setup the timer.  Honestly my only concern is the `venc/kvenc` naming.  It makes sense now but maybe extra comments to remind anyone new that this is our key/regular video encoding additions.

**Comment by colhom** [comment]
haha I thought a game dev would appreciate that. I worry about the threading overhead, we're doing 4 thread spawn/joins per frame. So 4 threads/frame * 10 fps * 20 captures = 800 threads / second. That's a little high... but want to take some time to get the worker queue implementation right so it's this way for now. Apparently reasonably modern linux kernels can easily do thousands of threads per second, so not ideal but should be fine for now :tm:

**Comment by colhom** [comment]
> I love that you are using the signal/slot interface to setup the timer. Honestly my only concern is the venc/kvenc naming

that edit distance is definitely way too close

**Comment by delaneyj** [comment]
> haha I thought a game dev would appreciate that. I worry about the threading overhead, we're doing 4 thread spawn/joins per frame. So 4 threads/frame * 10 fps * 20 captures = 800 threads / second. That's a little high... but want to take some time to get the worker queue implementation right so it's this way for now. Apparently reasonably modern linux kernels can easily do thousands of threads per second, so not ideal but should be fine for now

Normally I'd agree but my guess is the mem->GPU bandwidth will be the limiting factor vs the use of threads as signals.  If it becomes a problem then that's a good problem to have.

File: scripts/rvnbag_play.py [hunk 1]
```diff
@@ -271,37 +229,36 @@ def __init__( self, bagfile, loop):
...
 
-        try:
-            msg = msg_cont.deserialize()
-        except Exception as e:
-            self._set_terminate( ExitStatus.BAD_DESERIAL, "Unable to publish msg to '{}' topic, cannot deserialize: {}".format(msg_cont.topic,e))
+        if msg["topic"] not in self._all_publishers:
+            self._create_publisher(msg["topic"],msg["raw_msg"][-1])
 
-        self._all_publishers[msg_cont.topic].publish(msg)
+        pmsg = rospy.msg.AnyMsg()
```
Comment (colhom) [inline scripts/rvnbag_play.py:259]: Want to call out this bit for future reference, as it took me awhile to figure out how to use the `AnyMsg` class.

It allows you to read and publish a message without de-serializing it, or even having access to it's class definition. This is probably the single biggest source of CPU saving in this PR, as `rvnbag_play.py` was previously deserializing every message (some messages can be > 100mb) before publishing.

File: scripts/rvnbag_play.py [hunk 2]
```diff
@@ -271,37 +229,36 @@ def __init__( self, bagfile, loop):
...
-                msg_cont.topic,
-                msg_cont.msg_class,
-                latch      = self.bag_reader.is_latched(msg_cont.topic),
-                queue_size = 1000 ) # RVN::FIX: queue_size=???
-
+                topic,
+                msg_class,
+                latch      = self.bag_reader.is_latched(topic),
+                queue_size = 16,
+                tcp_nodelay = True,
```
Comment (colhom) [inline scripts/rvnbag_play.py:242]: Another obscure call-out here. This disables a network efficiency optimization when sending ros messages on the wire, which has the downside introducing more variable delay to message publishing.

Comment (colhom) [reply scripts/rvnbag_play.py:242]: By disabling it, we're trading decreased network/bandwidth efficiency for lower, less variable delays when sending messages.

Comment (colhom) [reply scripts/rvnbag_play.py:242]: If `tcp_nodelay` is left `False`, then [Nagle's algorithm](https://searchnetworking.techtarget.com/definition/Nagles-algorithm) is used to batch smaller message buffers together in a single TCP tx buffer ( and possibly as well on the rx buffer on the rviz side, would have to read up more)

**Comment by colhom** [comment]
>  If it becomes a problem then that's a good problem to have.

I think it eventually will, some of the beefier ec2 instances have 4 or more teslas in them, and more than enough CPU to shuffle buffers around. 

What we really want are [go rountines in C++ AKA boost fibers](https://www.boost.org/doc/libs/1_67_0/libs/fiber/doc/html/index.html), so we can keep our nice end-one-end implicit locking scheme and just change the threads to fibers- done.

Pre-req for that is pulling all our stuff out into a separate Catkin package and linking to a newer version of boost. The rest of ROS needs boost 1.5.whatever specifically, which doesn't have fibers.

**Pull request merged**